### PR TITLE
[SPARK-30072][SQL] Create dedicated planner for subqueries

### PR DESCRIPTION
**Pull request by dbaliafroozeh**
### What changes were proposed in this pull request?

This PR changes subquery planning by calling the planner and plan preparation rules on the subquery plan directly. Before we were creating a `QueryExecution` instance for subqueries to get the executedPlan. This would re-run analysis and optimization on the subqueries plan. Running the analysis again on an optimized query plan can have unwanted consequences, as some rules, for example `DecimalPrecision`, are not idempotent.

As an example, consider the expression `1.7 * avg(a)` which after applying the `DecimalPrecision` rule becomes:

```
promote_precision(1.7) * promote_precision(avg(a))
```

After the optimization, more specifically the constant folding rule, this expression becomes:

```
1.7 * promote_precision(avg(a))
```

Now if we run the analyzer on this optimized query again, we will get:

```
promote_precision(1.7) * promote_precision(promote_precision(avg(a)))
```

Which will later optimized as:

```
1.7 * promote_precision(promote_precision(avg(a)))
```

As can be seen, re-running the analysis and optimization on this expression results in an expression with extra nested promote_preceision nodes. Adding unneeded nodes to the plan is problematic because it can eliminate situations where we can reuse the plan. 

We opted to introduce dedicated planners for subuqueries, instead of making the DecimalPrecision rule idempotent, because this eliminates this entire category of problems. Another benefit is that planning time for subqueries is reduced.

### How was this patch tested?
Unit tests

### CODE REVIEW

**Comment by hvanhovell** [comment]
ok to test

**Review by hvanhovell — APPROVED** [review]
LGTM - pending jenkins

**Comment by dongjoon-hyun** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 1]
```diff
@@ -81,10 +81,8 @@ class QueryExecution(
   }
 
   lazy val sparkPlan: SparkPlan = tracker.measurePhase(QueryPlanningTracker.PLANNING) {
-    SparkSession.setActiveSession(sparkSession)
-    // TODO: We use next(), i.e. take the first plan returned by the planner, here for now,
-    //       but we will implement to choose the best plan.
     // Clone the logical plan here, in case the planner rules change the states of the logical plan.
+    QueryExecution.createSparkPlan(sparkSession, planner, optimizedPlan.clone())
     planner.plan(ReturnAnswer(optimizedPlan.clone())).next()
```
Comment (peter-toth) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: Shouldn't we remove this line?

Comment (dbaliafroozeh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: Thanks, good catch, missed removing this line.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala [hunk 2]
```diff
@@ -113,7 +112,8 @@ case class InsertAdaptiveSparkPlan(
     val queryExec = new QueryExecution(session, plan)
```
Comment (peter-toth) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala]: And this one too?

**Comment by hvanhovell** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/dynamicpruning/PlanDynamicPruningFilters.scala [hunk 3]
```diff
@@ -55,22 +55,24 @@ case class PlanDynamicPruningFilters(sparkSession: SparkSession)
...
             case BroadcastHashJoinExec(_, _, _, BuildLeft, _, left, _) =>
-              left.sameResult(qe.sparkPlan)
+              left.sameResult(sparkPlan)
             case BroadcastHashJoinExec(_, _, _, BuildRight, _, _, right) =>
-              right.sameResult(qe.sparkPlan)
+              right.sameResult(sparkPlan)
             case _ => false
           }.isDefined
 
+        val executedPlan = QueryExecution.planSubquery(sparkSession, sparkPlan)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/dynamicpruning/PlanDynamicPruningFilters.scala]: Move this into line 74?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 4]
```diff
@@ -248,3 +227,70 @@ class QueryExecution(
...
 }
+
+object QueryExecution {
+  /**
+   * Construct a sequence of rules that are used to prepare a planned [[SparkPlan]] for execution.
+   * These rules will make sure subqueries are planned, make use the data partitioning and ordering
+   * are correct, insert whole stage code gen, and try to reduce the work done by reusing exchanges
+   * and subqueries.
+   */
+  private[execution] def preparations(sparkSession: SparkSession): Seq[Rule[SparkPlan]] = {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: nit: we don't need outside `{` ... `}`

**Review by maropu — APPROVED** [review]
Ah, I see. The change looks reasonable. LGTM except for the minor comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 5]
```diff
@@ -131,11 +131,10 @@ case class AdaptiveSparkPlanExec(
       // Make sure we only update Spark UI if this plan's `QueryExecution` object matches the one
```
Comment (peter-toth) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala]: I think this comment should be amended a bit.

**Comment by peter-toth** [comment]
Just minor comments but LGTM.

**Comment by dbaliafroozeh** [comment]
@peter-toth @maropu thanks, I've addressed the minor issues, could you please do a final check and if it's good merge it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 6]
```diff
@@ -248,3 +227,67 @@ class QueryExecution(
...
+    //       but we will implement to choose the best plan.
+    planner.plan(ReturnAnswer(plan)).next()
+  }
+
+  /**
+   * Plan a subquery. Prepare the [[SparkPlan]] for execution.
+   */
+  def planSubquery(spark: SparkSession, plan: SparkPlan): SparkPlan = {
+    prepareForExecution(preparations(spark), plan)
+  }
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: nit: I know that this method for now is only called on subquery's plan. But it actually doesn't impose that, just plans any SparkPlan. Maybe rename to prepareExecutePlan.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 7]
```diff
@@ -248,3 +227,67 @@ class QueryExecution(
...
+
+  /**
+   * Plan a subquery. Transform the subquery's [[LogicalPlan]] into a [[SparkPlan]] and prepare the
+   * [[SparkPlan]] for execution.
+   */
+  def planSubquery(spark: SparkSession, plan: LogicalPlan): SparkPlan = {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: nit: ditto. maybe rename to prepareSparkPlan.

**Review by viirya — APPROVED** [review]
few minor comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 8]
```diff
@@ -45,11 +45,18 @@ case class ReturnAnswer(child: LogicalPlan) extends UnaryNode {
...
+ * @param correlated flag that indicates the subquery is correlated, and will be rewritten into a
+ *                   join during analysis.
  */
-case class Subquery(child: LogicalPlan) extends OrderPreservingUnaryNode {
+case class Subquery(child: LogicalPlan, correlated: Boolean) extends OrderPreservingUnaryNode {
   override def output: Seq[Attribute] = child.output
 }
 
+object Subquery {
+  def fromExpression(s: SubqueryExpression): Subquery = Subquery(s.plan, s.children.nonEmpty)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: We can reuse SubqueryExpression.hasCorrelatedSubquery.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala (hunk 6 above)
Comment (dbaliafroozeh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: Sounds reasonable to me, @hvanhovell what do you think about renaming it?

Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: Makes sense.

Comment (dbaliafroozeh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: I'm renaming both these methods to `prepareExecutedPlan` as one of them prepares a `SparkPlan` for execution and the other one a `LogicalPlan`. I think having the same method name makes sense for these methods.

**Comment by hvanhovell** [comment]
Merging to master

**Pull request closed without merging**